pkg/restore/systable_restore.go

```diff
@@ -0,0 +1,163 @@
+		return
+	}
+	db, ok := rc.getDatabaseByName(sysDB)
+	if !ok {
+		// Or should we create the database here?
+		log.Warn("target database not exist, aborting", zap.String("database", sysDB))
+		return
+	}
+
+	tablesRestored := make([]string, 0, len(originDatabase.Tables))
+	for _, table := range originDatabase.Tables {
+		tableName := table.Info.Name
+		if f.MatchTable(sysDB, tableName.O) {
+			rc.replaceTemporaryTableToSystable(ctx, tableName.L, db)
+		}
+		tablesRestored = append(tablesRestored, tableName.L)
+	}
+	if err := rc.afterSystemTablesReplaced(ctx, tablesRestored); err != nil {
+		for _, e := range multierr.Errors(err) {
+			logutil.WarnTerm("error during winding up the restoration of system table", zap.String("database", sysDB), logutil.ShortError(e))
+		}
+	}
+}
+
+// database is a record of a database.
+// For fast querying whether a table exists and the temporary database of it.
+type database struct {
+	ExistingTables map[string]*model.TableInfo
+	Name           model.CIStr
+	TemporaryName  model.CIStr
+}
+
+// getDatabaseByName make a record of a database from info schema by its name.
+func (rc *Client) getDatabaseByName(name string) (*database, bool) {
+	infoSchema := rc.dom.InfoSchema()
+	schema, ok := infoSchema.SchemaByName(model.NewCIStr(name))
+	if !ok {
+		return nil, false
+	}
+	db := &database{
+		ExistingTables: map[string]*model.TableInfo{},
+		Name:           model.NewCIStr(name),
+		TemporaryName:  utils.TemporaryDBName(name),
+	}
+	for _, t := range schema.Tables {
+		db.ExistingTables[t.Name.L] = t
+	}
+	return db, true
+}
+
+// afterSystemTablesReplaced do some extra work for special system tables.
+// e.g. after inserting to the table mysql.user, we must execute `FLUSH PRIVILEGES` to allow it take effect.
+func (rc *Client) afterSystemTablesReplaced(ctx context.Context, tables []string) error {
+	needFlushStat := false
+	var err error
+	for _, table := range tables {
+		switch {
+		case table == "user":
+			// We cannot execute `rc.dom.NotifyUpdatePrivilege` here, because there isn't
+			// sessionctx.Context provided by the glue.
+			// TODO: update the glue type and allow we retrive a session context from it.
+			err = multierr.Append(err, errors.Annotatef(berrors.ErrUnsupportedSystemTable,
+				"restored user info may not take effect, until you should execute `FLUSH PRIVILEGES` manually"))
+		case strings.HasPrefix(table, "stats_"):
+			needFlushStat = true
+		}
+	}
+	if needFlushStat {
+		// The newly created tables have different table IDs with original tables,
+		// 	hence the old statistics are invalid.
+		//
+		// TODO:
+		// 	Plan A) rewrite the IDs in the `rc.statsHandler.Update(rc.dom.InfoSchema())`.
+		//  Plan B) give the user a SQL to let him update it manually.
+		//  If plan A applied, we can deprecate the origin interface for backing up statistics.
+		err = multierr.Append(err, errors.Annotatef(berrors.ErrUnsupportedSystemTable,
```

Comment:
hold on, this error cannot be emitted *after* that `REPLACE INTO` stuff has taken place, this will break the stats of other existing tables.

Reply:
Yep. 🤔

Reply:
Done